File: src/frontend/src/lib/utils/mcpConfig.ts
/**
 * The user's trusted-MCP-server configuration, persisted on-chain (keyed by
 * anchor) so it syncs across all of the identity's devices — unlike the
 * CLI-access toggle, which is device-local. It is written via `mcp_set_config`,
 * authenticated as the identity, so only the user (never a page that initiates
 * a connect request) can change it. The `/mcp` connect flow reads it as the
 * source of truth at connect time.
 *
 * The config has two parts:
 *  - `enabled`: the feature's master toggle for this identity.
 *  - `url`: the trusted server URL, kept verbatim (so the Settings UI can probe
 *    a path-based endpoint like `https://host/mcp`); trust matching is by origin.
 *
 * It reaches the frontend two ways: certified, as `IdentityInfo.mcp_config` on
 * the `identity_info` update call, and uncertified, from the `mcp_get_config`
 * query below.
 */
import type { ActorSubclass } from "@icp-sdk/core/agent";
import type {
  _SERVICE,
  McpConfig as CanisterMcpConfig,
} from "$lib/generated/internet_identity_types";

export interface McpConfig {
  /** Master toggle for the feature on this identity. */
  enabled: boolean;
  /** The trusted server URL, or undefined when none is set. */
  url: string | undefined;
}

// Candid `opt text` -> `string | undefined`.
const fromOpt = (opt: [] | [string]): string | undefined =>
  opt.length === 0 ? undefined : opt[0];

/**
 * Decode a candid `opt McpConfig` — `IdentityInfo.mcp_config` or a
 * `mcp_get_config` reply. `undefined` when the identity never wrote a config.
 */
export const fromCanisterMcpConfig = (
  config: [] | [CanisterMcpConfig],
): McpConfig | undefined =>
  config.length === 0
    ? undefined
    : { enabled: config[0].enabled, url: fromOpt(config[0].url) };

/** Origin (scheme + host[:port], no path) of a URL, or undefined if unparsable. */
export const originOf = (url: string): string | undefined => {
  try {
    return new URL(url).origin;
  } catch {
    return undefined;
  }
};

/**
 * Read the identity's synced MCP config with the `mcp_get_config` query.
 *
 * The reply is **uncertified** — a single malicious node can forge it — so it
 * may only steer UX that a certified value decides for real: the `/mcp`
 * pre-check and the untrusted screen's auto-advance, both of which end in
 * `mcpAuthorize` gating delivery on `prepare`'s certified `trusted_url`. Never
 * render trust from it, and never feed it into an `mcp_set_config` write: use
 * `IdentityInfo.mcp_config`, certified by the `identity_info` update call.
 */
export const readMcpConfig = async (
  actor: ActorSubclass<_SERVICE>,
  identityNumber: bigint,
): Promise<McpConfig | undefined> =>
  fromCanisterMcpConfig(await actor.mcp_get_config(identityNumber));

File: src/internet_identity/src/storage/storable/mcp_config.rs
use ic_stable_structures::storable::Bound;
use ic_stable_structures::Storable;
use minicbor::{Decode, Encode};
use std::borrow::Cow;

/// Per-anchor trusted-MCP-server configuration, persisted in its own stable map
/// (see [`crate::storage`]) so it syncs across all of the identity's devices —
/// unlike the CLI-access toggle, which is device-local. It is written via the
/// authenticated `mcp_set_config` method and read by the `/mcp` connect flow
/// (verify-at-connect) and the Settings UI via `mcp_get_config`. Kept separate
/// from the core anchor record so it never touches anchor serialization.
///
/// `Default` is the disabled, no-server state, returned for anchors that have
/// never written a config.
///
/// It also carries the anchor's current in-flight *registration delegation*
/// ([`pending_registration`](StorableMcpConfig::pending_registration)) — the
/// per-anchor pointer that keeps the registration index bounded to one entry
/// per anchor. This rides the config entry (which already exists per anchor)
/// rather than a separate map, so it adds no new per-anchor growth surface.
#[derive(Encode, Decode, Clone, Default, Debug, Eq, PartialEq)]
#[cbor(map)]
pub struct StorableMcpConfig {
    /// Master toggle for the feature on this identity.
    #[n(0)]
    pub enabled: bool,
    /// The trusted server URL, kept verbatim so the Settings UI can display and
    /// re-probe a path-based endpoint like `https://host/mcp`. Trust matching in
    /// the connect flow is by origin. `None` when no server is set.
    #[n(1)]
    pub url: Option<String>,
    /// Raw principal bytes of the MCP session currently registered for this
    /// identity (the key of the grant map, see [`crate::storage`]), or `None`
    /// when no session is registered. The grant map is keyed by principal, so
    /// this forward pointer is what lets a config change revoke the active
    /// session in O(1) — and what enforces at most one session per identity:
    /// `mcp_register_v2` replaces the previous grant through it.
    #[cbor(n(2), with = "minicbor::bytes")]
    pub session_principal: Option<Vec<u8>>,
    /// Raw principal bytes of the anchor's current in-flight registration
    /// delegation `P_reg` (the key of its registration-index entry), or `None`
    /// when none is pending. An anchor trusts one server and holds one session,
    /// so at most one registration is in flight: a new
    /// `prepare_mcp_registration_delegation` supersedes the previous one,
    /// evicting the entry this points at before minting the next — which keeps
    /// the registration index bounded to one entry per anchor (a single actor
    /// can't flood it). A config written before this field existed decodes with
    /// it set to `None` (the `#[cbor(map)]` derive maps an absent key 3 to
    /// `None`).
    #[cbor(n(3), with = "minicbor::bytes")]
    pub pending_registration: Option<Vec<u8>>,
}

impl Storable for StorableMcpConfig {
    fn to_bytes(&self) -> Cow<'_, [u8]> {
        let mut buffer = Vec::new();
        minicbor::encode(self, &mut buffer).expect("failed to encode StorableMcpConfig");
        Cow::Owned(buffer)
    }

    fn from_bytes(bytes: Cow<'_, [u8]>) -> Self {
        minicbor::decode(&bytes).expect("failed to decode StorableMcpConfig")
    }

    const BOUND: Bound = Bound::Unbounded;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_roundtrip_through_storable() {
        let config = StorableMcpConfig {
            enabled: true,
            url: Some("https://mcp.example.com/mcp".to_string()),
            session_principal: Some(vec![1, 2, 3, 4]),
            pending_registration: Some(vec![7; 29]),
        };
        let decoded = StorableMcpConfig::from_bytes(config.to_bytes());
        assert_eq!(decoded, config);
    }

    /// Configs written before the `session_principal` field existed are CBOR
    /// maps without key 2 — they must decode with the field `None`.
    #[test]
    fn should_decode_config_without_session_principal() {
        // Encode the pre-`session_principal` shape: a CBOR map with keys 0..=1.
        let mut buffer = Vec::new();
        let mut encoder = minicbor::Encoder::new(&mut buffer);
        encoder.map(2).unwrap();
        encoder.u8(0).unwrap().bool(true).unwrap();
        encoder
            .u8(1)
            .unwrap()
            .str("https://mcp.example.com/mcp")
            .unwrap();

        let decoded = StorableMcpConfig::from_bytes(Cow::Borrowed(&buffer));
        assert!(decoded.enabled);
        assert_eq!(decoded.url, Some("https://mcp.example.com/mcp".to_string()));
        assert_eq!(decoded.session_principal, None);
        // A field added later (key 3) is likewise absent -> None.
        assert_eq!(decoded.pending_registration, None);
    }

    /// Configs written before the `pending_registration` field existed are
    /// CBOR maps without key 3 — they must decode with the field `None`.
    #[test]
    fn should_decode_config_without_pending_registration() {
        // Encode the pre-`pending_registration` shape: a map with keys 0..=2.
        let mut buffer = Vec::new();
        let mut encoder = minicbor::Encoder::new(&mut buffer);
        encoder.map(3).unwrap();
        encoder.u8(0).unwrap().bool(true).unwrap();
        encoder
            .u8(1)
            .unwrap()
            .str("https://mcp.example.com/mcp")
            .unwrap();
        encoder.u8(2).unwrap().bytes(&[5; 29]).unwrap();

        let decoded = StorableMcpConfig::from_bytes(Cow::Borrowed(&buffer));
        assert!(decoded.enabled);
        assert_eq!(decoded.session_principal, Some(vec![5; 29]));
        assert_eq!(decoded.pending_registration, None);
    }

    /// Rollback safety: bytes written by the new encoder (with the later keys
    /// present) must remain decodable when the unknown keys are skipped —
    /// minicbor's `#[cbor(map)]` derive ignores unknown map keys, which is what
    /// a rolled-back (older) wasm relies on. Cheaply approximated here by
    /// decoding new-encoder bytes and checking the pre-existing fields survive.
    #[test]
    fn should_keep_preexisting_fields_when_later_keys_present() {
        let config = StorableMcpConfig {
            enabled: true,
            url: Some("https://mcp.example.com/mcp".to_string()),
            session_principal: Some(vec![5; 29]),
            pending_registration: Some(vec![6; 29]),
        };
        let decoded = StorableMcpConfig::from_bytes(config.to_bytes());
        assert!(decoded.enabled);
        assert_eq!(decoded.url, config.url);
        assert_eq!(decoded.session_principal, Some(vec![5; 29]));
    }
}
